keras_cv/losses/focal.py

```diff
@@ -0,0 +1,109 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+import tensorflow.keras.backend as K
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class FocalLoss(tf.losses.Loss):
+    """Implements Focal loss
+
+    Focal loss is a modified cross-entropy designed to perform better with
+    class imbalance. For this reason, it's commonly used with object detectors.
+
+    Args:
+        alpha: a float value between 0 and 1 representing a weighting factor
+            used to deal with class imbalance. Positive classes and negative
+            classes have alpha and (1 - alpha) as their weighting factors
+            respectively. Defaults to 0.25.
+        gamma: a positive float value representing the tunable focusing
+            parameter. Defaults to 2.
+        from_logits: Whether `y_pred` is expected to be a logits tensor. By
+            default, `y_pred` is assumed to encode a probability distribution.
+            Default to `False`.
+        label_smoothing: Float in [0, 1]. When > 0, label values are smoothed,
+            meaning the confidence on label values are relaxed. For example, if
+            `0.1`, use `0.1 / num_classes` for non-target labels and
+            `0.9 + 0.1 / num_classes` for target labels. Default to `0.0`.
+
+    References:
+        [Focal Loss paper](https://arxiv.org/abs/1708.02002)
+
+    Standalone usage:
+    ```python
+        y_true = tf.random.uniform([10], 0, maxval=4)
+        y_pred = tf.random.uniform([10], 0, maxval=4)
+        loss = FocalLoss()
+        loss(y_true, y_pred).numpy()
+    ```
+    Usage with the `compile()` API:
+    ```python
+        model.compile(optimizer='adam', loss=keras_cv.losses.FocalLoss())
+    ```
+    """
+
+    def __init__(
+        self,
+        alpha=0.25,
+        gamma=2,
+        from_logits=False,
+        label_smoothing=0,
+        reduction="none",
+        name="FocalLoss",
+    ):
+        super().__init__(reduction=reduction, name=name)
+        self._alpha = float(alpha)
+        self._gamma = float(gamma)
+        self.from_logits = from_logits
+        self.label_smoothing = label_smoothing
+
+    def _smooth_labels(self, y_true, y_pred):
+        num_classes = tf.cast(tf.shape(y_true)[-1], y_pred.dtype)
+        return y_true * (1.0 - self.label_smoothing) + (
+            self.label_smoothing / num_classes
+        )
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.convert_to_tensor(y_pred)
+        y_true = tf.cast(y_true, y_pred.dtype)
+
+        if self.from_logits:
```

Comment:
can you replace this with:

```
if self.from_logits:
    y_pred = tf.nn.sigmoid(y_pred)
```

and remove the probs variable entirely?

Reply:
ah, nevermind I see that line 91 re-uses this.  But perhaps it would be more performant to use the probabilities directly since we already compute them with sigmoid here?